### PR TITLE
[O2B-1140] Use observable data for detector provider

### PR DESCRIPTION
**Pull request by martinboulais**
#### I have a JIRA ticket
- [X] branch and/or PR name(s) include(s) JIRA ID
- [X] issue has "Fix version" assigned
- [X] issue "Status" is set to "In review"
- [X] PR labels are selected

Notable changes for developers:
- Use observable data for detectors provider

### CODE REVIEW

File: lib/public/views/Runs/Details/RunDetailsModel.js [hunk 1]
```diff
@@ -41,7 +41,8 @@ export class RunDetailsModel extends Observable {
         this._runId = null;
         this._runDetails = RemoteData.notAsked();
         this._eorReasonTypes = RemoteData.notAsked();
-        this._allDetectors = RemoteData.notAsked();
+        this._allDetectors$ = detectorsProvider.all$;
+        this._allDetectors$.bubbleTo(this);
```
Comment (xsalonx) [inline lib/public/views/Runs/Details/RunDetailsModel.js:45]: maybe also `this._detectors$` like in other classes, I mean up to you, I just don't like this name

File: lib/public/utilities/ObservableData.js [hunk 2]
```diff
@@ -129,9 +146,15 @@ class ObservableDataBuilder {
      * @return {ObservableData} the build observable data
      */
     build() {
-        const observableData = new ObservableDataWithProcessing(this._initialValue, this._processors);
+        const initialValue = this._initialValue ?? this._source?.getCurrent();
+        const observableData = new ObservableDataWithProcessing(initialValue, this._processors);
+        if (this._source) {
+            const source = this._source;
+            source?.observe(() => observableData.setCurrent(source.getCurrent.bind(source)()));
```
Comment (xsalonx) [inline lib/public/utilities/ObservableData.js:153]:
```suggestion
            source.observe(() => observableData.setCurrent(source.getCurrent()));
```

File: lib/public/utilities/ObservableData.js [hunk 3]
```diff
@@ -111,6 +117,17 @@ class ObservableDataBuilder {
         return this;
     }
 
+    /**
+     * Any time the source data value change it will be reflected to the created observable data
```
Comment (xsalonx) [inline lib/public/utilities/ObservableData.js:121]: I think `reflected` is quite too ambiguous. 
Maybe:
Current value is updated with source's one when changed

File: lib/public/services/detectors/detectorsProvider.js [hunk 4]
```diff
@@ -43,14 +52,54 @@ export class DetectorsProvider extends Observable {
...
+    get all$() {
+        if (this._isStale()) {
+            this._load();
+        }
+        return this._all$;
+    }
+
+    /**
+     * Return physical (meaning actual detectors, for example excluding TST) detectors list observable data
+     *
```
Comment (xsalonx) [inline lib/public/services/detectors/detectorsProvider.js:68]: I would remove `for example`, as it's precisely: `excluding TST` and nothing more for now

Comment (xsalonx) [reply lib/public/services/detectors/detectorsProvider.js:68]: ok, maybe, as I see my previous docs had same notion

File: lib/public/services/detectors/detectorsProvider.js [hunk 5]
```diff
@@ -43,14 +52,54 @@ export class DetectorsProvider extends Observable {
...
+    /**
+     * States if the cached detectors list need to be created or updated
+     *
+     * @return {boolean} true if the detectors list must be refreshed
+     * @private
+     */
+    _isStale() {
+        return this._all$.getCurrent().match({
+            NotAsked: () => true,
+            Other: () => false,
```
Comment (xsalonx) [inline lib/public/services/detectors/detectorsProvider.js:87]: No reload if failed?

**Review by xsalonx — CHANGES_REQUESTED** [review]

File: lib/public/services/detectors/detectorsProvider.js (hunk 5 above)
Comment (martinboulais) [reply lib/public/services/detectors/detectorsProvider.js:87]: For now I just put the same behavior, refactoring only

File: lib/public/views/Runs/RunPerPeriod/RunsPerLhcPeriodOverviewModel.js [hunk 6]
```diff
@@ -65,11 +46,13 @@ export class RunsPerLhcPeriodOverviewModel extends RunsOverviewModel {
...
-            runQualities: 'good',
-            definitions: 'PHYSICS',
-        } });
+        return buildUrl(endpint, {
+            filter: {
+                lhcPeriods: this._lhcPeriodName,
+                runQualities: 'good',
+                definitions: 'PHYSICS',
+            },
+        });
```
[the remark this answers is not in the document]
Comment (martinboulais) [reply lib/public/views/Runs/RunPerPeriod/RunsPerLhcPeriodOverviewModel.js:55]: It's reformat not refactoring, but if you really want I can revert

File: lib/public/views/Runs/Details/RunDetailsModel.js (hunk 1 above)
Comment (martinboulais) [reply lib/public/views/Runs/Details/RunDetailsModel.js:45]: Renamed, but not the getter to be clear from outter context that we are not getting specific detector list

File: lib/public/utilities/ObservableData.js (hunk 2 above)
Comment (martinboulais) [reply lib/public/utilities/ObservableData.js:153]: Removed the spurious `?` but the bind is mandatory here

File: lib/public/services/detectors/detectorsProvider.js (hunk 4 above)
Comment (martinboulais) [reply lib/public/services/detectors/detectorsProvider.js:68]: I think it's better to keep as an example to hide the actual implementation, maybe at some point we'll have other tests detectors? As you like, I was simply rephrasing your comment

File: lib/public/services/detectors/detectorsProvider.js (hunk 5 above)
Comment (martinboulais) [reply lib/public/services/detectors/detectorsProvider.js:87]: Simply refactoring, not changing the behavior

File: lib/public/views/Runs/RunPerPeriod/RunsPerLhcPeriodOverviewModel.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (xsalonx) [reply lib/public/views/Runs/RunPerPeriod/RunsPerLhcPeriodOverviewModel.js]: I was joking

File: lib/public/utilities/ObservableData.js (hunk 2 above)
Comment (martinboulais) [reply lib/public/utilities/ObservableData.js]: Removed, you are right

**Review by xsalonx — APPROVED** [review]

**Pull request merged**